projects/ngx-formentry/src/form-entry/directives/patient-identifier.directive.ts

```diff
@@ -0,0 +1,107 @@
+import { Directive, Input } from '@angular/core';
+import {
+  AbstractControl,
+  AsyncValidator,
+  NG_ASYNC_VALIDATORS,
+  ValidationErrors
+} from '@angular/forms';
+import { Observable, of } from 'rxjs';
+import {
+  catchError,
+  debounceTime,
+  distinctUntilChanged,
+  map,
+  switchMap
+} from 'rxjs/operators';
+import { HttpClient, HttpParams } from '@angular/common/http';
+import { LeafNode } from '../form-factory/form-node';
+import { Identifier } from '../value-adapters/types';
+import { Messages } from '../utils/messages';
+
+@Directive({
+  selector: '[ofePatientIdentifierValidator]',
+  providers: [
+    {
+      provide: NG_ASYNC_VALIDATORS,
+      useExisting: PatientIdentifierValidatorDirective,
+      multi: true
+    }
+  ]
+})
+export class PatientIdentifierValidatorDirective implements AsyncValidator {
+  @Input('ofePatientIdentifierValidator') node: LeafNode;
+  @Input() currentPatientId?: string;
+
+  constructor(private http: HttpClient) {}
+
+  validate(
+    control: AbstractControl
+  ): Promise<ValidationErrors | null> | Observable<ValidationErrors | null> {
+    const identifier = control.value;
+    const currentPatientExistingIdentifier: Array<Identifier> =
+      this.node.form.valueProcessingInfo?.patientIdentifiers ?? [];
+
+    if (
+      currentPatientExistingIdentifier.some(
+        (id) => id.identifier === identifier
+      )
+    ) {
+      // Disable the control, since the identifier is already assigned to the patient
+      control.disable();
+      return of(null);
+    }
+
+    // If the identifier is less than 3 characters, no need to validate
+    if (!identifier || identifier.length < 3) {
+      return of(null);
+    }
+
+    return of(identifier).pipe(
+      debounceTime(500),
+      distinctUntilChanged(),
+      switchMap((id) => this.validateIdentifier(id)),
+      map((response) => {
+        if (
+          response.isAssigned &&
+          response.patientId !== this.currentPatientId
+        ) {
+          return {
+            identifierTaken: {
+              message: Messages.identifierTaken
+            }
+          };
+        }
+        return null;
+      }),
+      catchError(() =>
+        of({
+          identifierError: {
+            message: Messages.identifierError
+          }
+        })
+      )
+    );
+  }
+
+  private validateIdentifier(identifier: string): Observable<any> {
+    // For testing purposes, change openmrsBase to the OpenMRS server you are using
+    const baseUrl = window?.['openmrsBase'] + '/ws/rest/v1' + '/';
+    const apiUrl = `${baseUrl}patient`;
+
+    const params = new HttpParams().set('q', identifier);
```

Comment:
I don't love this mechanism for checking if an ID is duplicated. We really need to use information about the _type_ of identifier being configured. Specifically, OpenMRS IDs, depending on identifier type can be:

* Unique
* Unique _per location_
* Allow duplicates

Reply:
I think doing this is a great idea, but rather than doing some query of the backend, we should just add a backend endpoint that can tell us if a specified identifier is valid for an identifier type. That also allows us to check _other_ validation constraints that this mechanism can't support. It wouldn't be too hard to write a REST endpoint that takes a (potential) identifier and identifier type and runs the validation and returns any errors that occur.